build-tools/src/main/kotlin/org/jetbrains/kotlin/ExecutorService.kt

```diff
@@ -311,3 +320,238 @@ private fun sshExecutor(project: Project) : ExecutorService = object : ExecutorS
+        // probably the companion should be started and connected before
+        for (i in 0..3) {
+            project.exec {
+                it.commandLine(idb, "list-targets", "--json")
+                it.standardOutput = out
+            }.assertNormalExitValue()
+            if (out.toString().trim().isNotEmpty()) break
+        }
+        return out.toString().run {
+            check(isNotEmpty())
+            @Serializable
+            data class DeviceTarget(val name: String, val udid: String, val state: String, val type: String)
+            split("\n")
+                    .filter { it.isNotEmpty() }
+                    .map { Json(strictMode = false).parse(DeviceTarget.serializer(), it) }
+                    .first {
+                        it.type == "device" && deviceName?.run { this == it.name } ?: true
+                    }
+                    .udid
+        }
+    }
+
+    private fun install(udid: String, bundlePath: String) {
+        val out = ByteArrayOutputStream()
+
+        val result = project.exec {
+            it.workingDir = xcProject.toFile()
+            it.commandLine = listOf(idb, "install", "--udid", udid, bundlePath)
+            it.standardOutput = out
+            it.errorOutput = out
+            it.isIgnoreExitValue = true
+        }
+        println(out.toString())
+        check(result.exitValue == 0) { "Installation of $bundlePath failed: $out" }
+    }
+
+    private fun uninstall(udid: String, bundleId: String) {
+        val out = ByteArrayOutputStream()
+
+        project.exec {
+            it.workingDir = xcProject.toFile()
+            it.commandLine = listOf(idb, "uninstall", "--udid", udid, bundleId)
+            it.standardOutput = out
+            it.errorOutput = out
+            it.isIgnoreExitValue = true
+        }
+        println(out.toString())
+    }
+
+    private fun startDebugServer(udid: String, bundleId: String): String {
+        val out = ByteArrayOutputStream()
+
+        val result = project.exec {
+            it.workingDir = xcProject.toFile()
+            it.commandLine = listOf(idb, "debugserver", "start", "--udid", udid, bundleId)
+            it.standardOutput = out
+            it.errorOutput = out
+        }
+        check(result.exitValue == 0) { "Failed to start debug server: $out" }
+        return out.toString()
+    }
+}
+
+fun prepareLauncher(project: Project): Copy = project.tasks.create("prepareLauncher", Copy::class.java).apply {
+    from(project.file("iosLauncher"))
+    into(Paths.get(project.testOutputRoot, "launcher"))
+}
+
+fun KonanTestExecutable.xcodeBuild() {
+    this.doBeforeRun = Action {
+        val signIdentity = project.findProperty("sign_identity") as? String ?: "iPhone Developer"
+        val developmentTeam = project.findProperty("development_team") as? String
+        val xcProject = Paths.get(project.testOutputRoot, "launcher")
+
+        val shellScript: String = // language=Bash
```

Comment:
I think this code may be cleaner.
1. Use raw string literals to avoid escaping `"`.
2. `listOf()` instead of `it += `

Reply:
1. But it still required to escape `$` that escapes via `${"$"}` that isn't much better.
2. You still have to do `it += listOf()` in the when.  I'll join the first cp and += dSym part